File: java/TestPriorityQueue.java
import java.util.PriorityQueue;

public class TestPriorityQueue {
  private static class Pair implements Comparable<Pair> {
    int first;
    int second;
    Pair(int a, int b) {
      first = a;
      second = b;
    }

    @Override
    public int compareTo(Pair a1) {
      return Double.compare((double)first / second,
                            (double)a1.first / a1.second);
    }
  }
  public static void testPriorityQueue() {
    PriorityQueue<Pair> q = new PriorityQueue<>();
    int[] a = {1, 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47};

    int n = a.length;
    for (int i = 0; i < n; i++) {
      for (int j = i + 1; j < n; j++) {
        Pair p = new Pair(a[i], a[j]);
        q.offer(p);
      }
    }

    int cnt = 0;
    while (!q.isEmpty()) {
      Pair p = q.poll();
      System.out.printf("%4d/%-4d  ", p.first, p.second);
      cnt++;
      if (cnt == 10) {
        System.out.printf("\n");
        cnt = 0;
      }
    }
  }

  public static void main(String[] args) {
	  testPriorityQueue();
  }
}

File: cpp/testPriorityQueue.cpp
#include <queue>
#include <utility>
#include <cstdio>
#include <functional>

namespace std
{
template <>
struct less<std::pair<int, int>>
{
    bool operator()(const std::pair<int, int>& a1, const std::pair<int, int> a2)
    {
        return (double)a1.first / a1.second < (double)a2.first / a2.second;
    }
};
}

int main()
{
    auto f = [](std::pair<int, int> a1, std::pair<int, int> a2) { return (double)a1.first / a1.second < double(a2.first) / a2.second; };

    std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, decltype(f)> q(f);
    // std::priority_queue<std::pair<int, int>,  decltype(f)> q(f);   //wrong
    // std::priority_queue<std::pair<int, int>> q;   //if we provide specialization of std::less, then we can use this form to construct a priority_queue


    int a[]{1, 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47};
    int n = sizeof(a) / sizeof(a[0]);

    for (int i = 0; i < n; i++)
    {
        for (int j = i + 1; j < n; j++)
        {
            auto p = std::make_pair(a[i], a[j]);
            q.push(p);
        }
    }

    int cnt = 0;
    while (!q.empty())
    {

        auto p = q.top();
        q.pop();

        printf("%4d/%-4d  ", p.first, p.second);
        cnt++;
        if (cnt == 10)
        {
            printf("\n");
            cnt = 0;
        }
    }

    return 0;
}
